File: typescript/src/provider/secretspec.ts
/* Copyright (c) 2025 Voxgig Ltd, MIT License */

import { ProviderSpec, Provider, SekretoError, envkey, nodemod } from './support'

/** SecretSpec — https://secretspec.dev
 *
 * A declaration plus a chain of its own backends: the same shape as
 * sekreto one level down. A project that has declared its secrets there
 * should not have to declare them again here, so this reads through the
 * `secretspec` CLI rather than reimplementing its resolution.
 *
 * SecretSpec audits every read and refuses without `--reason`; sekreto
 * sends `sekreto` unless configured otherwise. */
export function secretspecprovider(options?: {
  command?: string
  file?: string
  profile?: string
  backend?: string
  reason?: string
  prefix?: string
}): Provider {
  const opts = options || {}
  const command = opts.command || 'secretspec'

  return {
    lookup: (name: string) => {
      const key = envkey(name, opts.prefix)

      const args: string[] = []
      if (opts.file) {
        args.push('--file', opts.file)
      }
      args.push('get', key)
      if (opts.backend) {
        args.push('--provider', opts.backend)
      }
      if (opts.profile) {
        args.push('--profile', opts.profile)
      }
      args.push('--reason', opts.reason || 'sekreto')

      const { spawnSync } = nodemod<typeof import('node:child_process')>('node:child_process')
      const run = spawnSync(command, args, { encoding: 'utf8' })

      if (run.error) {
        throw new SekretoError('sekreto: cannot run ' + command + ': ' + run.error.message)
      }

      if (0 === run.status) {
        // The value and one newline, and nothing else.
        return run.stdout.replace(/\n$/, '')
      }

      const why = (run.stderr || '').trim()

      if (secretspecmiss(why, key)) {
        return undefined
      }

      throw new SekretoError('sekreto: secretspec error: ' + (why || 'exit ' + run.status))
    },
    describe: () => 'secretspec' + (opts.backend ? ':' + opts.backend : ''),
  }
}

/** Does this SecretSpec failure mean "no such secret" rather than "I
 * could not answer"?
 *
 * SecretSpec says `Secret 'API_TOKEN' not found` for both a name it does
 * not declare and one declared with no value, and both are misses: this
 * store does not hold it, so the chain carries on.
 *
 * MATCHED ON THE WHOLE PHRASE, NOT ON "not found". SecretSpec also says
 * `Provider backend 'keyring' not found`, which is a store that could
 * not answer at all - and reading that as a miss is the worst failure
 * this library has, because the chain then falls through to a weaker
 * store without saying so. The key is required to appear, so the two
 * cannot be confused. */
function secretspecmiss(why: string, key: string): boolean {
  return why.includes("Secret '" + key + "' not found")
}

// Registering at import is what makes this module's presence the only
// thing that decides whether the kind exists in a build.
import { register } from './Registry'

register({
  name: 'secretspec',
  needs: ['node:child_process'],
  define: (spec: ProviderSpec) =>
    secretspecprovider({
      command: spec.command,
      file: spec.file,
      profile: spec.profile,
      backend: spec.backend,
      reason: spec.reason,
      prefix: spec.prefix,
    }),
})

File: lua/src/sekreto/plugins/secretspec.lua
-- The secretspec plugin: SecretSpec, through its CLI. Needs a child
-- process, so it is not a built-in kind.
--
-- A port of typescript/plugins/secretspec.ts, which is canonical.

local err = require('sekreto.err')
local name = require('sekreto.name')
local providers = require('sekreto.providers')
local support = require('sekreto.plugins.support')
local net = require('sekreto.plugins.net')

local fail = err.fail
local nonempty = providers.nonempty
local providerplugin = providers.providerplugin
local first = support.first
local stripnewline = support.stripnewline
local runcmd = net.runcmd

local M = {}

--- Does this SecretSpec failure mean "no such secret" rather than "I
--- could not answer"?
---
--- MATCHED ON THE WHOLE PHRASE, NOT ON "not found". SecretSpec also says
--- `Provider backend 'keyring' not found`, which is a store that could
--- not answer at all - and reading that as a miss is the worst failure
--- this library has, because the chain then falls through to a weaker
--- store without saying so. The key is required to appear, so the two
--- cannot be confused.
function M.secretspecmiss(why, key)
  return nil ~= why:find("Secret '" .. key .. "' not found", 1, true)
end

--- SecretSpec (https://secretspec.dev), read through its CLI.
---
--- `backend` selects one of SecretSpec's own backends (`--provider`) and
--- is called `backend` here only because `provider` already means a
--- sekreto provider. A reason is required, not optional: SecretSpec
--- records every read in an audit log and refuses to read without one.
local function secretspec(command, file, profile, backend, reason, prefix)
  local usecommand = nonempty(command) and command or 'secretspec'

  return {
    lookup = function(secret)
      local key = name.envkey(secret, prefix)

      local argv = { usecommand }
      if nonempty(file) then
        argv[#argv + 1] = '--file'
        argv[#argv + 1] = file
      end
      argv[#argv + 1] = 'get'
      argv[#argv + 1] = key
      if nonempty(backend) then
        argv[#argv + 1] = '--provider'
        argv[#argv + 1] = backend
      end
      if nonempty(profile) then
        argv[#argv + 1] = '--profile'
        argv[#argv + 1] = profile
      end
      argv[#argv + 1] = '--reason'
      argv[#argv + 1] = first(reason, 'sekreto')

      local out, why, status = runcmd(argv, nil, usecommand)

      if 0 == status then
        return stripnewline(out)
      end

      if M.secretspecmiss(why, key) then
        return nil
      end

      fail('sekreto: secretspec error: ' ..
        (('' == why) and ('exit ' .. status) or why))
    end,
    describe = function()
      return 'secretspec' .. (nonempty(backend) and (':' .. backend) or '')
    end,
  }
end

--- The kind, as a voxgig/plugin definition.
M.secretspec = providerplugin('secretspec', function(spec)
  return secretspec(spec.command, spec.file, spec.profile, spec.backend,
    spec.reason, spec.prefix)
end)

return M
